File: shoring/desk/src/main/java/com/panopset/marin/bootstrap/VersionClassGenerator.java
package com.panopset.marin.bootstrap;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

import com.panopset.compat.Fileop;
import com.panopset.compat.Logop;
import com.panopset.compat.Stringop;
import com.panopset.flywheel.FlywheelBuilder;

public class VersionClassGenerator {

  private String versionString;  
  
  public VersionClassGenerator(String version) {
    versionString = version;
  }

  public static void main(String... args) throws IOException {
    new VersionClassGenerator(args == null ? "" : args.length < 1 ? "" : args[0]).updateVersion();
  }

  private void updateVersion() {
    updateAppVersionClass();
    updatePoms();
  }

  private void updatePoms() {
    updateProject("../shoring");
    updateProject("compat");
    updateProject("db");
    updateProject("blackjackEngine");
    updateProject("minpin");
    updateProject("lowerclass");
    updateProject("flywheel");
    updateProject("fxapp");
    updateProject("desk");
    updateProject("panopset");
    updateProject("web");
    updateProject("joist");
  }

  private void updateProject(String project) {
    String pp = String.format("../%s/pom.xml", project);
    updatePom(pp);
  }

  private void updatePom(String pp) {
    String fr = "<version>";
    String tm = "<version>%s</version>";
    String vr = getVersionString();
    replaceFirstLine(pp, fr, String.format(tm, vr));
  }

  private void replaceFirstLine(String path, String lineToReplaceContaining,
      String fullReplacementLine) {
    File file = new File(path);
    List<String> source = Fileop.readLines(file);
    List<String> strs = Stringop.replaceFirstLinePreserveIndentation(source,
        lineToReplaceContaining, fullReplacementLine);
    if (source.equals(strs)) {
      Logop.info(String.format("No changes to %s, skipping...", path));
    } else {
      Fileop.write(strs, file);
      Logop.info(String.format("%s, updated with %s.", path, fullReplacementLine));
    }
  }

  private void updateAppVersionClass() {
    File vf = new File("src/main/java/com/panopset/compat/AppVersion.java");
    if (!vf.getParentFile().exists()) {
      vf = new File("../compat/src/main/java/com/panopset/compat/AppVersion.java");
    }
    if (!vf.getParentFile().exists()) {
      Logop.error("Parent directory not found", vf);
      return;
    }
    String result = new FlywheelBuilder()
        .inputResourcePath(VersionClassGenerator.class,
            "/com/panopset/marin/bootstrap/version.tmplt")
        .map("version_timestamp", getVersionString()).construct().exec();
    if (Stringop.isPopulated(result)) {
      Fileop.write(result, vf);
    }
  }

  private String getVersionString() {
    if (Stringop.isBlank(versionString)) {
      versionString = new SimpleDateFormat("yyyy.MM.dd").format(new Date());
    }
    return versionString;
  }
}

File: code/shoring/desk/src/main/kotlin/com/panopset/marin/bootstrap/VersionClassGenerator.kt
package com.panopset.marin.bootstrap

import com.panopset.compat.Fileop
import com.panopset.compat.Logz
import com.panopset.compat.Stringop.isPopulated
import com.panopset.compat.Stringop.replaceFirstLinePreserveIndentation
import com.panopset.desk.DeployProperties
import com.panopset.flywheel.FlywheelBuilder
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

class VersionClassGenerator(private val srcDirectory: String, private var versionString: String) {
    private fun updateVersion() {
        if (updateAppVersionClass()) {
            updatePoms()
        }
    }

    private fun updatePoms() {
        updatePom("$srcDirectory/shoring/pom.xml")
        updateProject("compat")
        updateProject("blackjackEngine")
        updateProject("flywheel")
        updateProject("fxapp")
        updateProject("desk")
        updatePom("$srcDirectory/legacy/pom.xml")
    }

    private fun updateProject(project: String) {
        val pp = "$srcDirectory/shoring/$project/pom.xml"
        updatePom(pp)
    }

    private fun updatePom(pp: String) {
        println("Updating: $pp")
        val fr = "<version>"
        val tm = "<version>%s</version>"
        val vr = getVersionString()
        replaceFirstLine(pp, fr, String.format(tm, vr))
    }

    private fun replaceFirstLine(
        path: String, lineToReplaceContaining: String,
        fullReplacementLine: String
    ) {
        val file = File(path)
        val source = Fileop.readLines( file)
        val strs = replaceFirstLinePreserveIndentation(
            source,
            lineToReplaceContaining, fullReplacementLine
        )
        if (source == strs) {
            Logz.info(String.format("No changes to %s, skipping...", path))
        } else {
            Fileop.write( strs, file)
            Logz.info(String.format("%s, updated with %s.", path, fullReplacementLine))
        }
    }

    private fun updateAppVersionClass(): Boolean {
        val vf = File("$srcDirectory/shoring/compat/src/main/kotlin/com/panopset/compat/AppVersion.kt")
        if (!vf.parentFile.exists()) {
            Logz.errorMsg("Parent directory not found", vf)
            return false
        }
        val result = FlywheelBuilder()
            .scriptFilePath(
                "$srcDirectory/../docs/templates/version.tmplt"
            )
            .map("panopset_desk_version", getVersionString())
            .map("panopset_desk_build", getBuildString())
            .construct().exec()
        if (isPopulated(result)) {
            Fileop.write( result, vf)
        }
        return true
    }

    private fun getBuildString(): String {
        return SimpleDateFormat("yyyyMMddHHmm").format(Date())
    }

    private fun getVersionString(): String {
        return versionString
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            VersionClassGenerator(args[0], DeployProperties().getPanopsetVersion()).updateVersion()
        }
    }
}
